File: bbclient/src/com/bbclient/HazelcastDriver.java
package com.bbclient;

import java.util.concurrent.*;

class HazelcastDriver extends Thread implements AutoCloseable {
    final Logger logger;
    final ExecutorService pool;
    final Main.RunnablesList runnablesList;
    final long statsFrequencyMillis;

    /*
     * Synchronized
     */
    boolean drain = false;

    public synchronized boolean isDraining() {
        return this.drain;
    }

    public synchronized void setDrain() {
        this.drain = true;
    }

    /*
     * Constructor and non-synchronized
     */

    public HazelcastDriver(String name,
                           Main.RunnablesList runnablesList,
                           long statsFrequencyMillis) {
        this.logger = new Logger(name);
        this.runnablesList = runnablesList;
        this.statsFrequencyMillis = statsFrequencyMillis;

        final int numThreads = Main.localTestMode ? 2 : 6;
        final int numThreadsMax = (Main.localTestMode ? 2 : 8) * numThreads;
        final int threadQueueSize = numThreadsMax * 2;
        int keepAliveTimeSec = 60; // seconds

        this.pool = new ThreadPoolExecutor(numThreads, numThreadsMax,
                keepAliveTimeSec, TimeUnit.SECONDS,
                new ArrayBlockingQueue<>(threadQueueSize),
                new ThreadPoolExecutor.AbortPolicy());
    }

    public void submit(Runnable runnable) {
        boolean submitted = false;
        var threadpoolSubmitBackoffMillis = 250; // ms
        while (!isDraining() && !submitted) {
            try {
                pool.submit(runnable);
                submitted = true;
            } catch (RejectedExecutionException e) {
                try {
                    Thread.sleep(threadpoolSubmitBackoffMillis);
                } catch (InterruptedException e2) {
                    logger.log("*** RECEIVED INTERRUPTED EXCEPTION [SUBMIT] *** %s".formatted(e2));
                    Thread.currentThread().interrupt();
                }
            }
        }
    }

    public void run() {
        logger.log("Starting up operations");
        var statsStopwatch = new Stopwatch(statsFrequencyMillis);

        while (!isDraining()) {
            for (Runnable runnable : runnablesList) {
                if (isDraining())
                    break;
                submit(runnable);
                if (!isDraining() && statsStopwatch.isTimeOver()) {
                    logger.log("STATS => " + runnablesList.listRunnablesToStatsString());
                }
            }
        }
    }

    public boolean reachedMinimumStatsPopulation() {
        return this.runnablesList.stream().allMatch(IMapMethodRunnable::hasReachedMinimumPopulation);
    }

    private void drain() throws InterruptedException {
        setDrain();
        pool.shutdown();
        if (!pool.awaitTermination(2, TimeUnit.SECONDS)) {
            logger.log("Timed out waiting for termination in drain()");
            pool.shutdownNow();
        }
    }

    public void drainAndJoin() {
        try {
            drain();
            this.join();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    public String drainAndGetStats() {
        try {
            drain();
            return runnablesList.listRunnablesToCSV();
        } catch (InterruptedException e) {
            logger.log("*** RECEIVED EXCEPTION [GET] *** %s".formatted(e));
            throw new RuntimeException(e);
        }
    }

    public void close() {
        try {
            drain();
        } catch (InterruptedException e) {
            pool.shutdownNow();
            Thread.currentThread().interrupt();
        }
        pool.shutdownNow();

        for (var runnable: runnablesList) {
            runnable.close();
        }
    }
}

File: bbclient/src/com/bbclient/HazelcastDriver.kt
package com.bbclient

import kotlinx.coroutines.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import kotlin.time.Duration

class HazelcastDriver(
    name: String,
    private val runnablesList: RunnablesList,
    private val statsFrequency: Duration
) : AutoCloseable {
    val logger: Logger = Logger(name)

    private val driverScope = CoroutineScope(
        Job() +
                Dispatchers.Default +
                CoroutineName("Executor"))

    class DispatcherExecutor(private val maxConcurrency: Int): AutoCloseable {
        private var concurrency = AtomicInteger(0)

        private val executorScope = CoroutineScope(
            Job() +
                    Dispatchers.IO +
                    CoroutineName("DispatcherExecutor")
        )

        fun submit(f: suspend () -> Unit): Boolean {
            var submitted = false
            if (concurrency.incrementAndGet() <= maxConcurrency) {
                submitted = true
                executorScope.launch {
                    f()
                    concurrency.decrementAndGet()
                }
            } else {
                concurrency.decrementAndGet()
            }
            return submitted
        }

        override fun close() {
            runBlocking {
                executorScope.coroutineContext.job.cancelAndJoin()
                executorScope.coroutineContext.cancelChildren()
                executorScope.cancel()
            }
        }
    }

    private val maxConcurrency = if (Main.localTestMode) 16 else 32

    private val pool = DispatcherExecutor(maxConcurrency)

    private var isDraining = AtomicBoolean(false)

    private fun setDrain() {
        isDraining.set(true)
    }

    fun start() = driverScope.launch {
        logger.log("Starting up operations")
        val statsStopwatch = Stopwatch(statsFrequency)

        while (!isDraining.get()) {
            for (runnable in runnablesList) {
                var submitted = false
                while (!submitted && !isDraining.get()) {
                    submitted = pool.submit { runnable.run() }

                    if (!submitted)
                        delay(100)
                }

                if (statsStopwatch.isTimeOver() && !isDraining.get())
                        logger.log("STATS => " + runnablesList.listRunnablesToStatsString())
            }
        }
    }

    fun reachedMinimumStatsPopulation(): Boolean {
        return runnablesList.stream().allMatch { obj: IMapMethodRunnable -> obj.hasReachedMinimumPopulation() }
    }

    private fun drain() {
        runBlocking {
            setDrain()
            pool.close()
            driverScope.coroutineContext.job.cancelAndJoin()
            driverScope.cancel()
        }
    }

    fun drainAndGetStats(): String {
        drain()
        return runnablesList.listRunnablesToCSV()
    }

    override fun close() {
        drain()
        for (runnable in runnablesList) {
            runnable.close()
        }
    }
}
